src/attached_probe.cpp

```diff
@@ -524,78 +488,68 @@ bool AttachedProbe::resolve_offset_uprobe(bool safe_mode, bool has_multiple_aps)
-    if (sym.start) {
-      LOG(V1) << "vmlinux: using " << path;
-      return path;
-    }
-  }
-
-  return "";
-}
-
 void AttachedProbe::resolve_offset_kprobe(bool safe_mode)
 {
-  struct symbol sym = {};
-  std::string &symbol = probe_.attach_point;
-  uint64_t func_offset = probe_.func_offset;
-  offset_ = func_offset;
-
 #ifndef HAVE_UNSAFE_PROBE
   safe_mode = true;
 #endif
 
-  if (func_offset == 0)
+  offset_ = probe_.func_offset;
+
+  // If we are using only the symbol, we don't need to check the offset.
+  bool is_symbol_kprobe = !probe_.attach_point.empty();
+  if (is_symbol_kprobe && probe_.func_offset == 0)
     return;
 
-  sym.name = symbol;
-  const struct vmlinux_location *locs = vmlinux_locs;
-  struct vmlinux_location locs_env[] = {
-    { nullptr, false },
-    { nullptr, false },
-  };
-  char *env_path = std::getenv("BPFTRACE_VMLINUX");
-  if (env_path) {
-    locs_env[0].path = env_path;
-    locs = locs_env;
-  }
+  // Setup the symbol to resolve, either using the address or the name.
+  struct symbol sym = {};
+  if (is_symbol_kprobe)
+    sym.name = probe_.attach_point;
+  else
+    sym.address = probe_.address;
 
-  std::string path = find_vmlinux(locs, sym);
-  if (path.empty()) {
-    LOG(V1) << "Could not resolve symbol " << symbol
+  auto path = find_vmlinux(&sym);
+  if (!path.has_value()) {
+    if (probe_.attach_point.empty())
+      throw FatalUserException("Could not resolve address: " +
+                               std::to_string(probe_.address));
+
+    LOG(V1) << "Could not resolve symbol " << probe_.attach_point
             << ". Skipping usermode offset checking.";
     LOG(V1) << "The kernel will verify the safety of the location but "
                "will also allow the offset to be in a different symbol.";
-
     return;
   }
 
-  if (func_offset >= sym.size)
-    throw FatalUserException("Offset outside the function bounds ('" + symbol +
-                             "' size is " + std::to_string(sym.size) + ")");
+  // Populate probe_ fields according to the resolved symbol.
+  if (is_symbol_kprobe) {
+    probe_.address = sym.start + probe_.func_offset;
+  } else {
+    probe_.attach_point = std::move(sym.name);
+    if (__builtin_sub_overflow(probe_.address, sym.start, &probe_.func_offset))
+      throw FatalUserException("Offset before the function bounds ('" +
+                               probe_.attach_point + "' address is " +
+                               std::to_string(sym.start) + ")");
+    // Set the name of the probe to the resolved symbol+offset, so that failure
+    // to attach can be ignored if the user set ConfigMissingProbes::warn.
+    probe_.name = "kprobe:" + probe_.attach_point + "+" +
+                  std::to_string(probe_.func_offset);
+  }
 
-  uint64_t sym_offset = resolve_offset(path, probe_.attach_point, probe_.loc);
+  if (probe_.func_offset >= sym.size)
+    throw FatalUserException("Offset outside the function bounds ('" +
```

Comment:
same here

Reply:
I'm not sure it makes sense for both those cases to use the MissingProbes config.
1. The error above should be marked "unreachable". In `find_vmlinux(&sym)`, `libbcc` looks through each symbol and returns the one containing the given `probe_.address`. So we know that `probe_.address` is `>= sym.start`, otherwise we'd have failed at the `find_vmlinux` call. So this check is more about defensive programming.
2. The second one makes more sense to me, but we have to distinguish the case where the offset was give by the user or if it comes from multiple probes (like probing an inlined function). Moreover, it also need to be changed in `resolve_offset_uprobe`.
Let me know what you think and I'll update the PR accordingly 🙂

Reply:
For the error above #1, which you say is unreachable. Then it should probably just be a `LOG(BUG)` if we don't expect it to ever happen.

For #2. Yeah I agree. @ajor and I were sorta discussing similar around the alignment issue. We didn't really come to a consensus on how to handle the common case in terms of erroring or just warning and skipping attachment for the probe. We can push this to a discussion to a different PR though.